packages/react-dom/src/server/ReactDOMServerExternalRuntime.js

```diff
@@ -1,18 +1,164 @@
+ * @flow
+ */
+
+// Imports are resolved statically by the closure compiler in release bundles
+// and by rollup in jest unit tests
 import {
   clientRenderBoundary,
   completeBoundaryWithStyles,
   completeBoundary,
   completeSegment,
-} from 'react-dom-bindings/src/server/fizz-instruction-set/ReactDOMFizzInstructionSet';
-
-// Intentionally does nothing. Implementation will be added in future PR.
-// eslint-disable-next-line no-unused-vars
-const observer = new MutationObserver(mutations => {
-  // These are only called so I can check what the module output looks like. The
-  // code is unreachable.
-  clientRenderBoundary();
-  completeBoundaryWithStyles();
-  completeBoundary();
-  completeSegment();
-});
+} from '../../../react-dom-bindings/src/server/fizz-instruction-set/ReactDOMFizzInstructionSet';
+
+// This runtime may be sent to the client multiple times (if FizzServer.render
+//  is called more than once). Here, we check whether the mutation observer
+//  was already created / installed
+if (!window.reactInstructionObserver && window.MutationObserver) {
+  // TODO: eventually remove
+  // We currently need to set these globals for compatibility with
+  // ReactDOMFizzInstructionSet
+  window.$RC = completeBoundary;
+  window.$RM = new Map();
+  window.reactInstructionObserver = new MutationObserver(mutations => {
+    mutations.forEach(mutation => {
+      mutation.addedNodes.forEach(handleNode);
+    });
+  });
+
+  const existingNodes = document.getElementsByTagName('div');
+  for (let i = 0; i < existingNodes.length; i++) {
+    handleNode(existingNodes[i]);
+  }
+
+  window.reactInstructionObserver.observe(document.body, {
+    childList: true,
+    subtree: true,
+  });
+}
+
+function handleNode(node_ /*: Node */) {
+  if (node_.nodeType !== 1) {
+    return;
+  }
+  // $FlowFixMe[incompatible-type]
+  const node /*: HTMLElement*/ = node_;
+  const instr = node.getAttribute('data-react-instr');
+  if (!instr) {
+    return;
+  }
+
+  switch (instr) {
+    case '$RX':
+      clientRenderBoundaryImpl(
+        node.getAttribute('data-rarg0'),
+        node.getAttribute('data-rarg1'),
+        node.getAttribute('data-rarg2'),
+        node.getAttribute('data-rarg3'),
+      );
+      break;
+    case '$RR':
+      completeBoundaryWithStylesImpl(
+        node.getAttribute('data-rarg0'),
+        node.getAttribute('data-rarg1'),
+        node.getAttribute('data-rarg2'),
+      );
+      break;
+    case '$RC':
+      completeBoundaryImpl(
+        node.getAttribute('data-rarg0'),
+        node.getAttribute('data-rarg1'),
+        node.getAttribute('data-rarg2'),
+      );
+      break;
+    case '$RS':
+      completeSegmentImpl(
+        node.getAttribute('data-rarg0'),
+        node.getAttribute('data-rarg1'),
+      );
+      break;
+    default:
+      if (__DEV__) {
+        console.error('Unexpected instruction');
+      }
+  }
+}
+
+/// Typechecking code, follows from above functions
+function isSuspenseBoundaryID(s) {
```

Comment:
I'm not convinced all these dev-only runtime type checking code is worth it. We don't even have a dev build of this module. The only time it runs is when we run our Jest tests.